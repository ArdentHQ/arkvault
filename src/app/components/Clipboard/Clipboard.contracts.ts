import React from "react";
import { ButtonVariant } from "@/types";

interface ClipboardCommonProperties {
	data: string | object;
	options?: Record<string, any>;
	children: React.ReactNode;
	disabled?: boolean;
}

export type ClipboardButtonProperties = ClipboardCommonProperties & {
	variant: "button";
	buttonVariant?: ButtonVariant;
	wrapperClassName?: string;
	buttonClassName?: string;
} & React.ButtonHTMLAttributes<any>;

export type ClipboardIconProperties = ClipboardCommonProperties & {
	variant: "icon";
	tooltip?: string;
	tooltipDarkTheme?: boolean;
	iconButtonClassName?: string;
};

export type ClipboardProperties = ClipboardIconProperties | ClipboardButtonProperties;
