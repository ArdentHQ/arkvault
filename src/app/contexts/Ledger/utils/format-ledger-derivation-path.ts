import { LedgerDerivationScheme } from "@/app/contexts/Ledger/Ledger.contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
