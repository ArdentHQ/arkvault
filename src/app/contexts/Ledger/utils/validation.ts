import semver from "semver";
import { Coins } from "@payvo/sdk";
import { minVersionList } from "@/app/contexts/Ledger/Ledger.contracts";

export const hasRequiredAppVersion = async (coin: Coins.Coin) => {
	const coinId = coin.network().coin();

	if (minVersionList[coinId]) {
		const currentVersion = await coin.ledger().getVersion();

		if (semver.lt(currentVersion, minVersionList[coinId])) {
			return false;
		}
	}

	return true;
};
