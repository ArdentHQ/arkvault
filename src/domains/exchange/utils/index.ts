export * from "./assertions";
export * from "./errors";
