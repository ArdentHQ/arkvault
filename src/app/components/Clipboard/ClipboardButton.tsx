import cn from "classnames";
import { AnimatePresence, motion } from "framer-motion";
import React from "react";

import { ClipboardButtonProperties } from "./Clipboard.contracts";
import { getStyles } from "@/app/components/Button/Button.styles";
import { Icon } from "@/app/components/Icon";
import { useClipboard } from "@/app/hooks";
import { twMerge } from "tailwind-merge";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const ClipboardButton = ({
	buttonVariant = "secondary",
	data,
	options,
	wrapperClassName,
	buttonClassName,
	children,
	...properties
}: Omit<ClipboardButtonProperties, "variant">) => {
	const [hasCopied, copy] = useClipboard({
		resetAfter: 1000,
		...options,
	});

	return (
		<div className={cn("relative", wrapperClassName)}>
			<button
				type="button"
				onClick={() => copy(data)}
				data-testid="clipboard-button__wrapper"
				className={twMerge(getStyles({ variant: buttonVariant }), buttonClassName)}
				{...properties}
			>
				<div className="flex items-center space-x-2">{children}</div>
			</button>

			<AnimatePresence>
				{hasCopied && (
					<motion.div
						initial={{ opacity: 0 }}
						animate={{ opacity: 1, transition: { duration: 0.3 } }}
						exit={{ opacity: 0, transition: { duration: 0.3 } }}
						className={cn("absolute inset-0 flex items-center justify-center rounded", {
							"bg-theme-primary-100 dark:bg-theme-secondary-800": buttonVariant === "secondary",
							"bg-theme-primary-700": buttonVariant === "primary",
						})}
						data-testid="clipboard-button__checkmark"
					>
						<Icon
							name="Checkmark"
							className={cn({
								"text-theme-primary-600 dark:text-theme-secondary-200": buttonVariant === "secondary",
								"text-white": buttonVariant === "primary",
							})}
						/>
					</motion.div>
				)}
			</AnimatePresence>
		</div>
	);
};
