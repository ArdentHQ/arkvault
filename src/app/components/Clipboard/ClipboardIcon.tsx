import React from "react";
import { useTranslation } from "react-i18next";

import { Tooltip } from "@/app/components/Tooltip";
import { useClipboard } from "@/app/hooks";

import { ClipboardIconProperties } from "./Clipboard.contracts";

export const ClipboardIcon = ({
	data,
	tooltip,
	tooltipDarkTheme,
	options,
	children,
	disabled,
}: Omit<ClipboardIconProperties, "variant">) => {
	const { t } = useTranslation();

	const [hasCopied, copy] = useClipboard({
		resetAfter: 1000,
		...options,
	});

	return (
		<Tooltip
			content={hasCopied ? t("COMMON.CLIPBOARD.SUCCESS") : tooltip || t("COMMON.CLIPBOARD.TOOLTIP_TEXT")}
			hideOnClick={false}
			theme={tooltipDarkTheme ? "dark" : undefined}
			disabled={disabled}
		>
			<button
				type="button"
				data-testid="clipboard-icon__wrapper"
				className="ring-focus relative focus:outline-none"
				onClick={() => copy(data)}
				data-ring-focus-margin="-m-1"
			>
				{children}
			</button>
		</Tooltip>
	);
};
