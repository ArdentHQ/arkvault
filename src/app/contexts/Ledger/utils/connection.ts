import { Coins } from "@ardenthq/sdk";
import retry, { AbortError, Options } from "p-retry";

import { formatLedgerDerivationPath } from "./format-ledger-derivation-path";
import { hasRequiredAppVersion } from "./validation";

export const accessLedgerDevice = async (coin: Coins.Coin) => {
	try {
		await coin.__construct();
		await coin.ledger().connect();
	} catch (error) {
		// If the device is open, continue normally.
		// Can be triggered when the user retries ledger connection.
		if (error.message !== "The device is already open.") {
			throw error;
		}
	}
};

export const accessLedgerApp = async ({ coin }: { coin: Coins.Coin }) => {
	await accessLedgerDevice(coin);

	if (!(await hasRequiredAppVersion(coin))) {
		throw new Error("VERSION_ERROR");
	}

	// Ensure that the app is accessible.
	await coin.ledger().getPublicKey(
		formatLedgerDerivationPath({
			coinType: coin.config().get<number>("network.constants.slip44"),
		}),
	);
};

export const persistLedgerConnection = async ({
	coin,
	options,
	hasRequestedAbort,
}: {
	coin: Coins.Coin;
	options: Options;
	hasRequestedAbort: () => boolean;
}) => {
	const retryAccess: any = async (attempts: number) => {
		if (hasRequestedAbort() && attempts > 1) {
			throw new AbortError("CONNECTION_ERROR");
		}

		try {
			await accessLedgerApp({ coin });
		} catch (error) {
			// Abort on version error or continue retrying access.
			if (error.message === "VERSION_ERROR") {
				throw new AbortError("VERSION_ERROR");
			}

			throw error;
		}
	};

	await retry(retryAccess, options);
};
