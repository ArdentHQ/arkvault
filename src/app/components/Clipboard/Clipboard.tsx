import React from "react";

import { ClipboardProperties } from "./Clipboard.contracts";
import { ClipboardButton } from "./ClipboardButton";
import { ClipboardIcon } from "./ClipboardIcon";

const defaultProps = {
	options: {},
};

export const Clipboard: React.VFC<ClipboardProperties> = ({
	options = defaultProps.options,
	variant,
	...properties
}) => {
	if (!properties.children) {
		return <></>;
	}

	if (variant === "icon") {
		return (
			<ClipboardIcon options={options} {...properties}>
				{properties.children}
			</ClipboardIcon>
		);
	}

	return (
		<ClipboardButton options={options} {...properties}>
			{properties.children}
		</ClipboardButton>
	);
};
