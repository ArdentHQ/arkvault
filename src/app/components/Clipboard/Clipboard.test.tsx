import userEvent from "@testing-library/user-event";
import React from "react";

import { Clipboard } from "./Clipboard";
import { act, render, screen, waitFor } from "@/utils/testing-library";

type VariantType = "icon" | "button";

describe("Clipboard", () => {
	beforeAll(() => {
		(navigator as any).clipboard = {
			writeText: jest.fn().mockResolvedValue("test"),
		};
	});

	afterAll(() => {
		(navigator as any).clipboard.writeText.mockRestore();
	});

	it.each<VariantType>(["icon", "button"])("should not render without children in variant type '%s'", (variant) => {
		// @ts-ignore
		const { asFragment } = render(<Clipboard variant={variant} data="" />);

		expect(screen.queryByTestId(`clipboard-${variant}__wrapper`)).not.toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it.each<VariantType>(["icon", "button"])("should render with children in variant type '%s'", (variant) => {
		const { asFragment } = render(
			<Clipboard variant={variant} data="">
				<span>Hello!</span>
			</Clipboard>,
		);

		expect(screen.queryByTestId(`clipboard-${variant}__wrapper`)).toHaveTextContent("Hello!");
		expect(asFragment()).toMatchSnapshot();
	});

	it.each([
		["string", "test string"],
		["object", { hello: "world" }],
	])("should work with data with type '%s'", async (dataType, data) => {
		jest.useFakeTimers();

		const onError = jest.fn();

		render(
			<Clipboard variant="icon" data={data} options={{ onError, resetAfter: 1000 }}>
				<span>Hello!</span>
			</Clipboard>,
		);

		userEvent.click(screen.getByTestId("clipboard-icon__wrapper"));

		act(() => {
			jest.runOnlyPendingTimers();
		});

		await waitFor(() => expect(onError).not.toHaveBeenCalled());
	});

	describe("on success", () => {
		it.each<VariantType>(["icon", "button"])(
			"should execute the onSuccess callback if given in variant type '%s'",
			async (variant) => {
				const onSuccess = jest.fn();

				render(
					<Clipboard variant={variant} data="" options={{ onSuccess }}>
						<span>Hello!</span>
					</Clipboard>,
				);

				userEvent.click(screen.getByTestId(`clipboard-${variant}__wrapper`));

				await waitFor(() => expect(onSuccess).toHaveBeenCalledWith(""));
			},
		);

		it.each<VariantType>(["icon", "button"])(
			"should execute no callback if missing in variant type '%s'",
			async (variant) => {
				const onSuccess = jest.fn();

				render(
					<Clipboard variant={variant} data="">
						<span>Hello!</span>
					</Clipboard>,
				);

				userEvent.click(screen.getByTestId(`clipboard-${variant}__wrapper`));

				await waitFor(() => expect(onSuccess).not.toHaveBeenCalled());
			},
		);
	});

	describe("on error", () => {
		beforeAll(() => {
			(navigator as any).clipboard = {
				writeText: jest.fn().mockRejectedValue(new Error("writeText rejected.")),
			};
		});

		it.each<VariantType>(["icon", "button"])(
			"should execute the onError callback if given in variant type '%s'",
			async (variant) => {
				const onError = jest.fn();

				render(
					<Clipboard variant={variant} data="" options={{ onError }}>
						<span>Hello!</span>
					</Clipboard>,
				);

				userEvent.click(screen.getByTestId(`clipboard-${variant}__wrapper`));

				await waitFor(() => expect(onError).toHaveBeenCalledWith());
			},
		);

		it.each<VariantType>(["icon", "button"])(
			"should execute no callback if missing in variant type '%s'",
			async (variant) => {
				const onError = jest.fn();

				render(
					<Clipboard variant={variant} data="">
						<span>Hello!</span>
					</Clipboard>,
				);

				userEvent.click(screen.getByTestId(`clipboard-${variant}__wrapper`));

				await waitFor(() => expect(onError).not.toHaveBeenCalledWith());
			},
		);
	});
});
