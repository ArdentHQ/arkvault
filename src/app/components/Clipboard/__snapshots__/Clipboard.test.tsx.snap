// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Clipboard > should not render without children in variant type 'button' 1`] = `<DocumentFragment />`;

exports[`Clipboard > should not render without children in variant type 'icon' 1`] = `<DocumentFragment />`;

exports[`Clipboard > should render with children in variant type 'button' 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <button
      class="css-6nyfvq"
      data-testid="clipboard-button__wrapper"
      type="button"
    >
      <div
        class="flex items-center space-x-2"
      >
        <span>
          Hello!
        </span>
      </div>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Clipboard > should render with children in variant type 'icon' 1`] = `
<DocumentFragment>
  <button
    class="ring-focus relative focus:outline-none"
    data-ring-focus-margin="-m-1"
    data-testid="clipboard-icon__wrapper"
    type="button"
  >
    <span>
      Hello!
    </span>
  </button>
</DocumentFragment>
`;
